charon/src/bin/charon-driver/translate/translate_types.rs

```diff
@@ -779,7 +781,118 @@ impl<'tcx, 'ctx> ItemTransCtx<'tcx, 'ctx> {
     }
 }
 
+fn translate_variant_layout(
+    variant_layout: &r_abi::LayoutData<r_abi::FieldIdx, r_abi::VariantIdx>,
+) -> VariantLayout {
+    match &variant_layout.fields {
+        r_abi::FieldsShape::Primitive
+        | r_abi::FieldsShape::Union(_)
+        | r_abi::FieldsShape::Array { .. } => VariantLayout::default(),
+        r_abi::FieldsShape::Arbitrary { offsets, .. } => {
+            let mut v = Vector::with_capacity(offsets.len());
+            for (i, o) in offsets.iter_enumerated() {
+                // The fields should have a total increasing index, but to be
+                // absolutely sure, we insert them at the correct index.
+                v.insert(FieldId::from_usize(i.index()), Some(o.bytes()));
+            }
```

Comment:
You'll need the `TypeDecl` around to know how many variants/fields there are.
```suggestion
            // TODO: get the `fields: &Vector<FieldId, Field>` from the `TypeDecl`
            let mut v = fields.map_ref(|_| None);
            for (i, o) in offsets.iter_enumerated() {
                v[FieldId::from_usize(i.index())] = Some(o.bytes());
            }
```